File: aoc2025-zig/src/days/day05.zig
const std = @import("std");
const common = @import("../common.zig");
const parse = @import("../parse.zig");

const Range = struct {
    start: usize,
    end: usize,
};

fn parseRanges(lines: [][]const u8, allocator: std.mem.Allocator) !std.ArrayList(Range) {
    var ranges: std.ArrayList(Range) = .empty;
    for (lines) |line| {
        const dash = std.mem.indexOfScalar(u8, line, '-') orelse continue;
        const start = try std.fmt.parseInt(usize, line[0..dash], 10);
        const end = try std.fmt.parseInt(usize, line[dash + 1 ..], 10);
        try ranges.append(allocator, Range{ .start = start, .end = end });
    }
    return ranges;
}

pub fn part1(input: [][][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    var ranges = try parseRanges(input[0], allocator);
    defer ranges.deinit(allocator);

    var total: usize = 0;
    for (input[1]) |line| {
        const id = try std.fmt.parseInt(usize, line, 10);
        for (ranges.items) |r| {
            if (id >= r.start and id <= r.end) {
                total += 1;
                break;
            }
        }
    }
    return std.fmt.allocPrint(allocator, "{d}", .{total});
}

pub fn part2(input: [][][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    var ranges = try parseRanges(input[0], allocator);
    defer ranges.deinit(allocator);

    std.mem.sort(Range, ranges.items, {}, struct {
        fn lt(_: void, a: Range, b: Range) bool {
            return a.start < b.start;
        }
    }.lt);

    var i: usize = 0;
    while (i + 1 < ranges.items.len) {
        const a = ranges.items[i];
        const b = ranges.items[i + 1];
        if (b.start <= a.end) {
            ranges.items[i] = Range{ .start = a.start, .end = @max(a.end, b.end) };
            _ = ranges.orderedRemove(i + 1);
        } else {
            i += 1;
        }
    }

    var total: usize = 0;
    for (ranges.items) |r| {
        total += r.end - r.start + 1;
    }
    return std.fmt.allocPrint(allocator, "{d}", .{total});
}

pub fn run(io: std.Io, allocator: std.mem.Allocator) !void {
    const input = try parse.sections(io, allocator, 5, false);

    const t1s = std.Io.Timestamp.now(io, .awake);
    const p1 = try part1(input, allocator);
    const t1e = std.Io.Timestamp.now(io, .awake);

    const t2s = std.Io.Timestamp.now(io, .awake);
    const p2 = try part2(input, allocator);
    const t2e = std.Io.Timestamp.now(io, .awake);

    common.runDay(io, 5, p1, p2, t1s, t1e, t2s, t2e);
}

File: aoc2025-go/days/day05/day05.go
package day05

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"

	"aoc2025-go/common"
	"aoc2025-go/parse"
)

const IS_TEST = false

type Range struct {
	Start int
	End   int
}

func parseRanges(lines []string) []Range {
	ranges := make([]Range, 0, len(lines))
	for _, line := range lines {
		r := strings.SplitN(line, "-", 2)
		start, _ := strconv.Atoi(r[0])
		end, _ := strconv.Atoi(r[1])
		ranges = append(ranges, Range{start, end})
	}
	return ranges
}

func Part1(input [][]string) string {
	ranges := parseRanges(input[0])
	total := 0
	for _, line := range input[1] {
		id, _ := strconv.Atoi(line)
		for _, r := range ranges {
			if id >= r.Start && id <= r.End {
				total++
				break
			}
		}
	}
	return fmt.Sprintf("%d", total)
}

func Part2(input [][]string) string {
	ranges := parseRanges(input[0])
	sort.Slice(ranges, func(i, j int) bool {
		return ranges[i].Start < ranges[j].Start
	})

	i := 0
	for i+1 < len(ranges) {
		a, b := ranges[i], ranges[i+1]
		if b.Start <= a.End {
			if b.End > a.End {
				ranges[i].End = b.End
			}
			ranges = append(ranges[:i+1], ranges[i+2:]...)
		} else {
			i++
		}
	}

	total := 0
	for _, r := range ranges {
		total += r.End - r.Start + 1
	}
	return fmt.Sprintf("%d", total)
}

func Run() {
	input := parse.Sections(5, IS_TEST)

	t1s := time.Now()
	p1 := Part1(input)
	t1 := time.Since(t1s)

	t2s := time.Now()
	p2 := Part2(input)
	t2 := time.Since(t2s)

	common.RunDay(5, p1, p2, t1, t2)
}
